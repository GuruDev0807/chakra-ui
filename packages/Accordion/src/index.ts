export * from "./Accordion.base"
export * from "./Accordion"
export * from "./Accordion.hook"
