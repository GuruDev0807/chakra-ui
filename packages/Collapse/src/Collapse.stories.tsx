import * as React from "react";
import { storiesOf } from "@storybook/react";
import useCollapse from "./Collapse";

const stories = storiesOf("Collapse", module);

function Example() {
  const collapse = useCollapse({
    defaultIsOpen: true,
  });
  return (
    <>
      <button {...collapse.toggle}>Toggle</button>
      <div {...collapse.collapse}>
        Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus
        terry richardson ad squid. Nihil anim keffiyeh helvetica, craft beer
        labore wes anderson cred nesciunt sapiente ea proident.
      </div>
    </>
  );
}
stories.add("basic", () => <Example />);

// stories.add("changing static height", () => {
//   function Example() {
//     const [show, setShow] = React.useState(false);

//     const on = () => setShow(!show);

//     return (
//       <>
//         <Collapse startingHeight={20} isOpen={show}>
//           Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus
//           terry richardson ad squid. Nihil anim keffiyeh helvetica, craft beer
//           labore wes anderson cred nesciunt sapiente ea proident.
//         </Collapse>
//         <button onClick={handleToggle}>{show ? "Collapse" : "Expand"}</button>
//       </>
//     );
//   }

//   return <Example />;
// });
