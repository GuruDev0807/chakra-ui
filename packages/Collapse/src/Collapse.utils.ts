import * as React from "react";

export function getElementHeight(el: React.RefObject<HTMLElement>) {
  if (!el || !el.current) return "auto";
  return el.current.scrollHeight;
}

export const defaultTransitionStyles: React.CSSProperties = {
  transitionDuration: "250ms",
  transitionTimingFunction: "ease-in",
};

export function join(str: string | undefined) {
  if (str) {
    const styles = ["height"];
    styles.push(...str.split(", "));
    return styles.join(", ");
  }
  return "height";
}

interface TransitionOptions {
  collapseStyles?: React.CSSProperties;
  expandStyles?: React.CSSProperties;
}

export function makeTransitionStyles(options: TransitionOptions) {
  const {
    expandStyles = defaultTransitionStyles,
    collapseStyles = defaultTransitionStyles,
  } = options;
  return {
    expandStyles: {
      ...expandStyles,
      transitionProperty: join(expandStyles.transitionProperty),
    },
    collapseStyles: {
      ...collapseStyles,
      transitionProperty: join(collapseStyles.transitionProperty),
    },
  };
}
