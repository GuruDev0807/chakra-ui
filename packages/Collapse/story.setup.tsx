import theme from "@chakra-ui/preset-base";
import {
  ColorMode,
  ColorModeProvider,
  createThemeContext,
  CSSReset,
} from "@chakra-ui/system";
import * as React from "react";
import { Global } from "@emotion/core";
import { PortalManager } from "@chakra-ui/portal";

const [ThemeProvider, useTheme] = createThemeContext(theme);

const setup = (story: () => any) => (
  <PortalManager>
    <ThemeProvider>
      <ColorModeProvider>
        <CSSReset />
        <ColorMode />
        <Global
          styles={{
            "*": {
              fontFamily: "system-ui",
            },
          }}
        />
        {story()}
      </ColorModeProvider>
    </ThemeProvider>
  </PortalManager>
);

export { ThemeProvider, useTheme };
export default setup;
