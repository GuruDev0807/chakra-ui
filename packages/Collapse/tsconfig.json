{
  "extends": "../../tsconfig.build.json",
  "include": ["src", "types"],
  "exclude": ["tests", "story"]
}
