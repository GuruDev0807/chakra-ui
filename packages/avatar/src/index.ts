export * from "./_avatar"
export * from "./avatar-group"
