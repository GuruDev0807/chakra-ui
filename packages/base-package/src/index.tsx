const test = { a: 3, b: { c: 45 } };
const test2 = test?.b
