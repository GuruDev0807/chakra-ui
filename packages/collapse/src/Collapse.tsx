import { chakra, PropsOf } from "@chakra-ui/system"
import {
  Transition,
  TransitionProps,
  TransitionStyles,
} from "@chakra-ui/transition"
import { ariaAttr, cx, mergeRefs, __DEV__ } from "@chakra-ui/utils"
import { useRect } from "@reach/rect"
import * as React from "react"
import {
  cloneElement,
  forwardRef,
  useState,
  Children,
  useRef,
  Ref,
} from "react"

export type CollapseProps = PropsOf<typeof chakra.div> & {
  /**
   * If `true`, the content will be visible
   */
  isOpen?: boolean
  /**
   * The height you want the content in it's collapsed state.
   * @default 0
   */
  startingHeight?: number
  /**
   * Custom styles for the Transition component's appear, entered and exiting states
   */
  config?: TransitionProps["styles"]
  /**
   * If `true`, the opacity of the content will be animated
   * @default true
   */
  animateOpacity?: boolean
  /**
   * The CSS `transition-duration` (in ms) to apply for the collapse animation
   *
   * @default
   * 150
   */
  timeout?: number
  /**
   * The CSS `transition-timing-function` to apply for the collapse animation
   *
   * @default
   * "ease"
   */
  easing?: string
}

export const Collapse = forwardRef(
  (props: CollapseProps, forwardedRef: React.Ref<any>) => {
    const {
      isOpen,
      children,
      config,
      startingHeight = 0,
      animateOpacity = true,
      className,
      style: htmlStyle,
      timeout = 150,
      easing = "ease",
      ...rest
    } = props

    const getStr = (property: string) => `${property} ${timeout}ms ${easing}`

    const transition = `${getStr("height")}, ${getStr("opacity")}, ${getStr(
      "transform",
    )}`

    const [hidden, setHidden] = useState(true)

    type ChildElement = React.ReactElement<{ ref: Ref<any> }>

    let child = children

    if (typeof children === "string") {
      child = <div>{children}</div>
    }

    const _child = Children.only(child) as ChildElement

    const ref = useRef<HTMLDivElement>(null)

    const rect = useRect(ref, true)
    const height = rect?.height ?? 0

    const styles: TransitionStyles = {
      init: {
        height: startingHeight,
        opacity: startingHeight ? 1 : 0,
      },
      entered: {
        height,
        opacity: 1,
        transform: "translateY(0)",
      },
      exiting: {
        height: startingHeight,
        opacity: startingHeight ? 1 : 0,
        transform: startingHeight > 0 ? "translateY(0)" : "translateY(-0.5rem)",
      },
    }

    return (
      <Transition
        in={isOpen}
        styles={config || styles}
        onEntered={() => setHidden(false)}
        onExited={() => setHidden(true)}
        timeout={{ enter: 0, exit: timeout }}
        transition={transition}
        unmountOnExit={false}
      >
        {styles => (
          <chakra.div
            ref={forwardedRef}
            className={cx("chakra-collapse", className)}
            aria-hidden={ariaAttr(hidden)}
            {...rest}
            style={{
              ...styles,
              overflow: "hidden",
              opacity: animateOpacity ? styles.opacity : 1,
              willChange: "height, opacity, transform",
              ...htmlStyle,
            }}
          >
            {cloneElement(_child, {
              ref: mergeRefs(ref, _child.props.ref),
            })}
          </chakra.div>
        )}
      </Transition>
    )
  },
)

if (__DEV__) {
  Collapse.displayName = "Collapse"
}
