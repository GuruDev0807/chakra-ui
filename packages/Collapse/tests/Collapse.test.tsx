import "@testing-library/jest-dom/extend-expect"
// import userEvent from "@testing-library/user-event"
import React from "react"
import { render } from "../../../test/utils"
import { Collapse } from "../src/Collapse"

test("Collapse renders correctly", () => {
  const utils = render(<Collapse>sfsjdfkbfkjdfdjksf</Collapse>)
  expect(utils.asFragment()).toMatchSnapshot()
})
