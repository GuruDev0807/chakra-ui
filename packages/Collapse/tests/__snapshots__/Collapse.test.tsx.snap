// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse renders correctly 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    style="height: 0px; overflow: hidden; opacity: 0;"
  >
    <div>
      sfsjdfkbfkjdfdjksf
    </div>
  </div>
  <div
    class="__chakra--portal-manager"
  />
</DocumentFragment>
`;
